flink-runtime/src/main/java/org/apache/flink/runtime/state/RegisteredBackendStateMetaInfo.java

```diff
@@ -92,10 +94,9 @@ public boolean isCompatibleWith(RegisteredBackendStateMetaInfo<?, ?> other) {
 			return false;
 		}
 
-		return ((namespaceSerializer == null && other.namespaceSerializer == null)
-					|| namespaceSerializer == null || other.namespaceSerializer == null
-					|| namespaceSerializer.isCompatibleWith(other.namespaceSerializer))
-				&& stateSerializer.isCompatibleWith(other.stateSerializer);
+		return (stateSerializer.isCompatibleWith(other.stateSerializer)) &&
+				(namespaceSerializer.isCompatibleWith(other.namespaceSerializer)
+						|| other.namespaceSerializer instanceof MigrationNamespaceSerializerProxy);
```

Comment:
I would add a short comment that it's important to check for the `MigrationNamespaceSerializerProxy`. It could even warrant a test. I can see how someone removes this line in the future.